doc/internal/man3/ossl_param_bld_init.pod

```diff
@@ -0,0 +1,181 @@
+
+ossl_param_build_push_TYPE() are a series of functions which will create
+OSSL_PARAM objects of the specified size and correct type for the B<val>
+argument.
+B<val> is stored by value and an expression or auto variable can be used.
+
+ossl_param_build_push_BN() is a function that will create an OSSL_PARAM object
+that holds the specified BIGNUM B<bn>.
+If B<bn> is marked as being securely allocated, the
+ossl_param_build_set_private() function is called which allocates the
+resulting OSSL_PARAM array and associated storage securely.
+The B<bn> argument is stored by reference and the underlying BIGNUM object
+must exist until after ossl_param_build_to_param() has been called.
+
+ossl_param_build_push_utf8_string() is a function that will create an OSSL_PARAM
+object that references the UTF8 string specified by B<buf>.
+If the length of the string, B<bsize>, is zero then it will be calculated.
+The string that B<buf> points to is stored by reference and must remain in
+scope until after ossl_param_build_to_param() has been called.
+
+ossl_param_build_push_octet_string() is a function that will create an OSSL_PARAM
+object that references the octet string specified by B<buf> and <bsize>.
+The memory that B<buf> points to is stored by reference and must remain in
+scope until after ossl_param_build_to_param() has been called.
+
+ossl_param_build_push_utf8_ptr() is a function that will create an OSSL_PARAM
+object that references the UTF8 string specified by B<buf>.
+If the length of the string, B<bsize>, is zero then it will be calculated.
+The string B<buf> points to is stored by reference and must remain in
+scope until the OSSL_PARAM array is freed.
+
+ossl_param_build_push_octet_ptr() is a function that will create an OSSL_PARAM
+object that references the octet string specified by B<buf>.
+The memory B<buf> points to is stored by reference and must remain in
+scope until the OSSL_PARAM array is freed.
+
+=head1 RETURN VALUES
+
+ossl_param_build_to_param() and ossl_param_bld_to_param_ex() return the
+allocated OSSL_PARAM array, or NULL on error.
+
+All of the ossl_param_build_push_TYPE functions return 1 on success and 0
+on error.
+
+=head1 NOTES
+
+The constant B<OSSL_PARAM_BLD_MAX> specifies the maximum number of parameters
+that can be added.
+Exceeding this will result in the push functions returning errors.
+
+=head1 EXAMPLES
+
+Both examples creating an OSSL_PARAM array that contains an RSA key.
+For both, the predefined key variables are:
+
+    BIGNUM *p, *q;  /* both prime */
+    BIGNUM *n;      /* = p * q */
+    unsigned int e; /* exponent, usually 65537 */
+    BIGNUM *d;      /* e^-1 */
+
+=head2 Example 1
+
+This example shows how to create an OSSL_PARAM array that contains an RSA
+private key.
+Because it is a private key, it is allocated in private memory.
+
+    OSSL_PARAM_BLD bld;
+    OSSL_PARAM *params;
+
+    ossl_param_build_init(&bld);
+    ossl_param_build_set_private(&bld);
```

Comment:
Is the set_private() only ever done initially?
If so it might make more sense to do ossl_param_build_init_private instead maybe?

Reply:
Heh, I made argument that it could be given late...  https://github.com/openssl/openssl/pull/9305#discussion_r300833784

Reply:
Internally, it is called if a securely allocated BIGNUM is encountered.  Otherwise it can be called anytime up until the either of the `ossl_param_bld_to_param` calls.

It could be added as an argument to the `ossl_param_bld_to_param` calls I guess...

Reply:
You should be able to toggle set_private on *and off.* I strongly believe that this kind of "change important semantics via side-effect" is a HORRIBLE anti-pattern.  PLEASE do not do this.

Reply:
I've switched to allowing the private to be turned on and off and queried.
A secure BIGNUM still enables this flag, which I think is the correct behaviour.